File: Programmers/2/152996. 시소 짝꿍/시소 짝꿍.java
import java.util.*;

class Solution {
    public long solution(int[] weights) {
        Map<Integer, Long> count = new HashMap<>();
        long answer = 0;

        for (int w : weights) {
            count.put(w, count.getOrDefault(w, 0L) + 1);
        }

        for (long c : count.values()) {
            if (c >= 2) {
                answer += c * (c - 1) / 2;
            }
        }

        for (int w : count.keySet()) {
            long c = count.get(w);

            if (w * 2 % 3 == 0) {
                int t = w * 2 / 3;
                if (t != w && count.containsKey(t))
                    answer += c * count.get(t);
            }

            if (w * 2 % 4 == 0) {
                int t = w * 2 / 4;
                if (t != w && count.containsKey(t))
                    answer += c * count.get(t);
            }

            if (w * 3 % 4 == 0) {
                int t = w * 3 / 4;
                if (t != w && count.containsKey(t))
                    answer += c * count.get(t);
            }
        }

        return answer;
    }
}

File: Programmers/2/152996. 시소 짝꿍/시소 짝꿍.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

long long solution(vector<int> weights) {
    long long answer = 0;
    unordered_map<int, long long> cnt;
    
    for (int w : weights) cnt[w]++;
    
    for (auto& [w, c] : cnt)
    {
        if (c >= 2) answer += c * (c - 1) / 2;
    }
    
    for (auto& [w, c] : cnt)
    {
        if (w * 2 % 3 == 0 && cnt.count(w * 2 / 3))
            answer += c * cnt[w * 2 / 3];
        
        if (w * 2 % 4 == 0 && cnt.count(w * 2 / 4))
            answer += c * cnt[w * 2 / 4];
        
        if (w * 3 % 4 == 0 && cnt.count(w * 3 / 4))
            answer += c * cnt[w * 3 / 4];
    }
    
    return answer;
}